fix(dev-infra): convert commit SHAs and PR numbers in changelog to links

gkalpak opened the pull request:
Previously, the commit SHAs and PR numbers referenced in the generated `CHANGELOG.md` were not automatically converted to links in the GitHub UI (as happens for release notes and issue/PR comments). This made it less straight-forward for someone reading the changelog to get to the commit/PR corresponding to a change.

This PR updates the tooling that generates the changelog to convert the commit SHA and the corresponding PR number (referenced at the end of the commit message header) to links. It also updates existing entries in `CHANGELOG.md` to have links for commit SHAs and corresponding PR numbers.

devversion commented on dev-infra/release/notes/templates/changelog.ts:
```diff
@@ -20,8 +20,10 @@ _%>
 <%_
   for (const commit of group.commits) {
+    const commitShaMarkdown = commitToLink(commit);
+    const commitHeaderMarkdown = commit.header.replace(/\\(#(\\d+)\\)$/, (_, g) => \`(\${prToLink(+g)})\`);
```
Optional nit: What do you think of moving this into a separate function in the context, like `replaceCommitHeaderPullRequestNumber`? Just bringing this up because it would make the template less concerned with such transformations, easier to understand what happens here, and the `\\` escape would not be needed.

gkalpak replied:
Yeah, I was on the fence on that. It felt a little too specific to put in the context class, but also quite complex for the template.
Happy to move it to the context.

devversion replied:
yeah, this is tough 😄 I'll leave the decision to you. I'm okay with either way.

devversion reviewed (approved):
LGTM nice! one optional comment (that comes down to preference 😄 )

gkalpak committed:
docs: convert commit SHAs and PR numbers in changelog to links

Since the recent update in the changelog format, the commit SHAs and PR
numbers referenced there were not links, making it less straight-forward
to get to the commit/PR corresponding to a change. The changelog
generation tooling has been updated in a previous commit to output links
for future changelog entries.

This commit updates existing entries in `CHANGELOG.md` to have links for
commit SHAs and corresponding PR numbers.

gkalpak committed:
fix(dev-infra): convert commit SHAs and PR numbers to links when generating changelog

Previously, the commit SHAs and PR numbers referenced in the generated
`CHANGELOG.md` were not automatically converted to links in the GitHub
UI (as happens for release notes and issue/PR comments). This made it
less straight-forward for someone reading the changelog to get to the
commit/PR corresponding to a change.

This commit updates the tooling that generates the changelog to convert
the commit SHA and the corresponding PR number (referenced at the end of
the commit message header) to links.

gkalpak committed:
fixup! fix(dev-infra): convert commit SHAs and PR numbers to links when generating changelog
